How to start vLLM on Beverin (using Slurm, Container Engine and Netstack) (#468)
eth-cscs/cscs-docs: docs/tutorials/ml/vllm-sglang-amdgpus.md

[](){#software-ml-vllm-sglang-amdgpus}

# vLLM/SGLang on AMD GPUs Tutorial

This tutorial will guide you through the steps required to setup a vLLM or SGLang container to serve a LLM on AMD GPUs.

In this specific tutorial we are going to show how to:

- setup a vLLM/SGLang container for AMD GPUs
- launch a single node LLM inference instance
- launch a multi-node LLM inference instance (with Slurm)
- and test it

!!! warning "What this guide is NOT"
    This guide is not about how to configure correctly/at best vLLM or SGLang on Alps clusters.
    For more details about vLLM or SGLang parameters and best setups for a specific model have a look at <https://recipes.vllm.ai/> and <https://docs.sglang.io/>, respectively, while for what concerns the Alps cluster setup, you will have to explore documentation for the specific cluster, experiment and benchmark.

### Prerequisites

This tutorial assumes that you are able to access the cluster via SSH. To set up access to CSCS systems, follow the guide [here][ref-ssh], and read through the documentation about the [ML platform][ref-platform-mlp].

> In particular, this tutorial aims at the Beverin cluster, but it should be possible to adapt it with minor changes to other clusters as well.

Reviewer: ```suggestion
In particular, this tutorial targets the Beverin cluster, but it should be possible to adapt it with minor changes to other clusters as well.
```

### Get the vLLM/SGLang image

> In this tutorial we are going to use an official vLLM/SGLang image for ROCm systems. From early 2026 the AMD's Docker images [`rocm/vllm`](https://hub.docker.com/r/rocm/vllm) and [`rocm/sgl-dev`](https://hub.docker.com/r/rocm/sgl-dev) (and others) have been deprecated (see, e.g., [here](https://docs.vllm.ai/en/v0.19.1/getting_started/installation/gpu/#use-amds-docker-images-deprecated) and [here](https://lmsysorg.mintlify.app/docs/hardware-platforms/amd_gpu#install-using-docker-recommended)) in favour of [`vllm/vllm-openai-rocm`](https://hub.docker.com/r/vllm/vllm-openai-rocm) and [`lmsysorg/sglang`](https://hub.docker.com/r/lmsysorg/sglang/tags) images, respectively. The following steps will show how to use the new images.

Author: > [here](https://lmsysorg.mintlify.app/docs/hardware-platforms/amd_gpu#install-using-docker-recommended)

at this link I can't find where it says that sglang images are deprecated, actually it seems that also `rocm/sgl-dev` are getting updates...

what am I missing?

Reviewer 2: It doesn't say that they have been deprecated but it says which images to use (lmsysorg/sglang`) and it's not the one on the AMD website (there is nothing happening anymore as for the VLLM case).

Author: legit, thanks for the explanation

> In this tutorial we are going to use an official vLLM/SGLang image for ROCm systems. From early 2026 the AMD's Docker images [`rocm/vllm`](https://hub.docker.com/r/rocm/vllm) and [`rocm/sgl-dev`](https://hub.docker.com/r/rocm/sgl-dev) (and others) have been deprecated (see, e.g., [here](https://docs.vllm.ai/en/v0.19.1/getting_started/installation/gpu/#use-amds-docker-images-deprecated) and [here](https://lmsysorg.mintlify.app/docs/hardware-platforms/amd_gpu#install-using-docker-recommended)) in favour of [`vllm/vllm-openai-rocm`](https://hub.docker.com/r/vllm/vllm-openai-rocm) and [`lmsysorg/sglang`](https://hub.docker.com/r/lmsysorg/sglang/tags) images, respectively. The following steps will show how to use the new images.

Reviewer: ```suggestion
In this tutorial we are going to use an official vLLM/SGLang image for ROCm systems. From early 2026 AMD's Docker images [`rocm/vllm`](https://hub.docker.com/r/rocm/vllm) and [`rocm/sgl-dev`](https://hub.docker.com/r/rocm/sgl-dev) (and others) have been deprecated (see, e.g., [here](https://docs.vllm.ai/en/v0.19.1/getting_started/installation/gpu/#use-amds-docker-images-deprecated) and [here](https://lmsysorg.mintlify.app/docs/hardware-platforms/amd_gpu#install-using-docker-recommended)) in favour of [`vllm/vllm-openai-rocm`](https://hub.docker.com/r/vllm/vllm-openai-rocm) and [`lmsysorg/sglang`](https://hub.docker.com/r/lmsysorg/sglang/tags) images, respectively. The following steps will show how to use the new images.
```

> Beforehand, if you have not done already, let us create a directory to keep track of all images used with the [Container Engine][ref-container-engine]. Since container images are large files and the filesystem is a shared resource, we need to apply [best practices for LUSTRE][ref-guides-storage-lustre] so they are properly distributed across storage nodes.

Reviewer: ```suggestion
Beforehand, if you have not done already, let us create a directory to keep track of all images used with the [Container Engine][ref-container-engine]. Since container images are large files and the filesystem is a shared resource, we need to apply [best practices for Lustre][ref-guides-storage-lustre] so they are properly distributed across storage nodes.
```

```console title="Container image directory with recommended LUSTRE settings"
mkdir -p $SCRATCH/ce-images
lfs setstripe -E 4M -c 1 -E 64M -c 4 -E -1 -c -1 -S 4M $SCRATCH/ce-images
```

> Now we can pull the docker image in our container images directory just created. The following command allows to import a docker image as squashfs archive that can be used with the [Container Engine][ref-container-engine].

Reviewer: ```suggestion
Now we can pull the docker image in our container images directory just created. The following command allows to import a docker image as a squashfs archive that can be used with the [Container Engine][ref-container-engine].
```

=== "`vLLM`"
    !!! example ""
        ```console
        enroot import \
            -o $SCRATCH/ce-images/inference-rocm.sqfs \
            docker://vllm/vllm-openai-rocm:latest
        ```

=== "`SGLang`"

>     !!! example "This uses Podman to pull the image first, to avoid issues related to filesystem permissions when using Enroot directly with Docker Hub."

Author: We should unify this step, since in principle it should not be application dependent. I hadn't any problem with permissions, but since it is a problem you had, I don't see a reason to not do the same for both applications.

Reviewer 2: I would leave it with Enroot for the VLLM case, because it is simpler and I don't think one will run into the issue encountered with SGLang: the issue raises because the SGLang images are not done very nicely and contain this conceptual problem that leads to the permission issue besides that they contain an enormous amount of files from the builds which could be cleaned up in theory.

In particular, this tutorial makes use of [netstack][ref-ce-netstack-source] and related hooks for binding compatible network libraries (mainly `libfabric`, `CXI` and `aws-ofi-nccl`) inside the container.

Save this as a TOML file named `env-inference.toml` so that you can use it later.

```toml
image = "/capstor/scratch/cscs/<username>/ce-images/inference-rocm.sqfs"
writable = true
entrypoint = false

[annotations]
com.hooks.netstack.source = "artifact"
com.hooks.netstack.version = "26.07.1"
com.hooks.netstack.name = "gpu:rocm7,cxi:13.1.0,ofi:2.6.0,aws:1.20.0"

com.hooks.cxi.enabled = "true"
com.hooks.aws_ofi_nccl.enabled = "true"

[env]
# Note: this solves an hang happening at NCCL initialization time
HWLOC_COMPONENTS="-gl"

FI_PROVIDER="cxi"
FI_CXI_RX_MATCH_MODE="software"
FI_MR_CACHE_MONITOR="disabled"
```

### Launch a Single Node instance

At this point everything is ready, the container image needs just to be launched and vLLM or SGLang started.

The simplest is to launch a single GPU instance with

=== "`vLLM`"
    !!! example ""
        ```console

>         srun --environment ./env-inference.toml -pmi300 \

Reviewer: Optional, but `env-inference.toml` could go into `~/.edf` where it'd be auto-detected and referred to by `--environment=env-inference` (no `.toml`). No objection to leaving it with the explicit path as well.
